### PR TITLE
Feature/downsample

### PR DESCRIPTION
**Pull request by nickp60**
This adds the downsampling rule that takes config parameters for depth and replicate.  The replicate gets used to seed the downsampling. Depth is provided as the number of total reads desired.  This follows seqkit's recommendation to sample  a proportion rather than an exact number of reads in order to avoid reading the whole file into memory.  So this calculates the number of reads in the sample, calculates  a generous downsampling proportion, samples that proportion, and selects the exact number or reads with `seqkit head`.

### CODE REVIEW

File: workflow/rules/downsample.smk [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+if not os.path.exists("logs"):
+    os.makedirs("logs")
+
+
+localrules:
+    all,
+
+
+depths = config["depths"]
```
Comment (funnell) [inline workflow/rules/downsample.smk:17]: perhaps you could add these to the config schema?

File: workflow/rules/downsample.smk [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        echo "Insufficient depth to downsample to requested depth"
+        exit 1
+        fi
+
+        # calculate the proportion needed, but avoid rounding issues by doing the *1.2 factor
+        generous_proportion=$( echo "print($reads_per_file/$nreads * 1.2)" | python3)
+
+        # sample -p + head is seqkit's recommended approach for downsampling,
+        # as it avoids having to read the whole thing into memory when you
+        #  directly to a number sample using -n
```
Comment (funnell) [inline workflow/rules/downsample.smk:84]: typo?

File: workflow/rules/downsample.smk [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        fq_tmpfile={params.tmpdir}/tmp.fq.gz
+        seqkit sample --rand-seed {wildcards.rep} --threads {threads} {input.R1} \
+          -p $generous_proportion -o $fq_tmpfile
+        seqkit head  -n $reads_per_file -o {output.R1} $fq_tmpfile
+
+        seqkit sample --rand-seed {wildcards.rep} --threads {threads} {input.R2} \
+          -p $generous_proportion -o $fq_tmpfile
+        seqkit head -n $reads_per_file -o {output.R2} $fq_tmpfile
+
+        rm $fq_tmpfile
```
Comment (funnell) [inline workflow/rules/downsample.smk:95]: do you need to do this after the first seqkit head as well?

File: workflow/rules/downsample.smk [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        R2=config["R2"],
+        R1ds=downsampled_fastqs_r1,
+        R2ds=downsampled_fastqs_r2,
+    output:
+        stats="reports/downsample_stats_{sample}.tab",
+    container:
+        config["docker_seqkit"]
+    shell:
+        """
+    seqkit stats --basename --tabular  --out-file {output.stats} {input}
```
Comment (funnell) [inline workflow/rules/downsample.smk:111]: weird indenting

**Comment by funnell** [comment]
what do you think about putting the two main rules into a rule group? https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#defining-groups-for-execution

File: workflow/rules/downsample.smk (hunk 1 above)
Comment (nickp60) [reply workflow/rules/downsample.smk:17]: I wasn't sure, its really only relevant to this app

File: workflow/rules/downsample.smk (hunk 3 above)
Comment (nickp60) [reply workflow/rules/downsample.smk:95]: I have it overwrite the same temp file.  I can add the extra rm it if that makes it more clear?

File: workflow/rules/downsample.smk (hunk 4 above)
Comment (nickp60) [reply workflow/rules/downsample.smk:111]: That is weird.

Comment (nickp60) [reply workflow/rules/downsample.smk:111]: Its looks like this behavior was introduced in 0.8. Snakefmt with 0.6.1 makes it look as expected. Unfortunately if we use the superlinter we are stuck with this formatting.

**Comment by nickp60** [comment]
I haven't worked with job groups, and I am a bit hesitant to add additional complexity.  Since our primary use case is HPC with a shared file system I don't think we need to optimize.  Would totally make sense on AWS or something, to avoid transferring the fastqs to a new VM.  Feel free to add it as a milestone for version 2.0.

File: workflow/rules/downsample.smk (hunk 2 above)
Comment (nickp60) [reply workflow/rules/downsample.smk]: fixed

**Comment by funnell** [comment]
job groups aren't necessary, for our purposes they would just group rules into the same LSF job in order to reduce scheduler overhead. Could be helpful for when the cluster is busy.

Feel free to merge if you feel there's no issue with re-using seqkit temp files 👍

**Review by funnell — APPROVED** [review]

File: workflow/rules/downsample.smk (hunk 3 above)
Comment (nickp60) [reply workflow/rules/downsample.smk:95]: Sure.  Ive added the additional rm .

**Comment by nickp60** [comment]
I've added an issue for the job groups. https://github.com/vdblab/vdblab-shotgun/issues/39

**Pull request merged**